### ISSUE TITLE
Unable to reproduce reported performance

### ISSUE THREAD

**Issue by sebastienwood** [danieltan07/dagmm#3]
Hello,

thanks for the good work on this paper reproduction. I tried to run it a few time to have a better understanding of DAGMM. However, I can't manage to reproduce the performance reported in the readme/paper.

Methodology is as follow : 
- run in train mode with no pretrained arg
- switch arg "mode" to test, and arg "pretrained_model" to "200_194"

The run outputs are in this range (minor variations, but nowhere close to 95%+) : 
Threshold : -0.8012019991874695
Accuracy : 0.6735, Precision : 0.5066, Recall : 0.2911, F-score : 0.3698

I ran it 5 times (train + test) with no modification of the code. Using Pytorch 0.4 and Python 3.6. On the last run I tried to delete the folder "models" before running.

If anyone knows how to tackle this issue it would be of great help ! Thanks :)

**Comment by GwuieonJin** [comment]
i also suffer this problem. if anyone have some clue, please let us know. Thanks!

**Comment by sebastienwood** [comment]
With further tests I managed to make it work in the Jupyter notebook provided. I suspect some saving/loading model is the root cause of this issue.

**Comment by danieltan07** [comment]
Hi @sebastienwood , sorry for the late reply, I was swamped with deadlines and did not notice your previous message. To be honest, I am not sure what is causing the inconsistencies. I haven't had the time to dig further into this issue. You might be right about the saving and loading part since I generate / preprocess / save the data separately. I was also thinking of looking into the some implementation details that were not really detailed in the paper such as type of normalization, choice of epsilon, or the computation of the determinant (I used numpy's implementation but I am not sure if the gradients can/should back propagate through that). Let me know if you find some bugs or figure out what is causing the issue.

**Comment by ykwon0407** [comment]
Hello, @sebastienwood  ! Could you give us more details? I also have experienced this replication issue.

**Comment by sebastienwood** [comment]
The Jupyter Notebook provided in the repository works out of the box. You can start at the block : 

class hyperparams():
    def __init__(self, config):
        self.__dict__.update(**config)
defaults = {
    'lr' : 1e-4,
    'num_epochs' : 200,
    'batch_size' : 1024,
    'gmm_k' : 4,
    'lambda_energy' : 0.1,
    'lambda_cov_diag' : 0.005,
    'pretrained_model' : None,
    'mode' : 'train',
    'use_tensorboard' : False,
    'data_path' : 'kdd_cup.npz',

    'log_path' : './dagmm/logs',
    'model_save_path' : './dagmm/models',
    'sample_path' : './dagmm/samples',
    'test_sample_path' : './dagmm/test_samples',
    'result_path' : './dagmm/results',

    'log_step' : 194//4,
    'sample_step' : 194,
    'model_save_step' : 194,
}

(the previous blocks are for data preprocessing).

I believe the problem occurs in a python terminal at the moment when you try to load a pretrained model (cf. first post). You could try to rework the code a bit so that the train/test is executed sequentially rather than separated with a save/load operation.
Tell me if I can provide more details :)

**Comment by ykwon0407** [comment]
@sebastienwood  Thanks! Actually, I did it in a sequential fashion, but still, I have no idea :( 

Based on my experiments, even when I check through the notebook, I managed to achieve 0.9x F1 scores but not always. Sometimes the F1 was like around 0.3. (Maybe I am missing something trivial.)
